src/tdigest.c

```diff
@@ -0,0 +1,512 @@
+}
+
+TDigest *
+ts_tdigest_create(void)
+{
+	return ts_tdigest_create_with_compression(TDIGEST_DEFAULT_COMPRESSION);
+}
+
+TDigest *
+ts_tdigest_create_with_compression(int compression)
+{
+	uint32 size;
+	TDigest *t;
+
+	if (compression < TDIGEST_MIN_COMPRESSION || compression > TDIGEST_MAX_COMPRESSION)
+		elog(ERROR,
+			 "compression parameter should be in [%d, %d], got %d",
+			 TDIGEST_MIN_COMPRESSION,
+			 TDIGEST_MAX_COMPRESSION,
+			 compression);
+
+	size = ceil(compression * M_PI / 2) + 1;
+
+	/* palloc0 means no manual nil'ing/setting to 0 of values necessary */
+	t = palloc0(sizeof(TDigest) + size * sizeof(Centroid));
+
+	t->compression = 1.0 * compression;
+	t->threshold = estimate_compression_threshold(compression);
+	t->size = size;
+	t->min = INFINITY;
+
+	SET_VARSIZE(t, ts_tdigest_size(t));
+
+	return t;
+}
+
+void
+ts_tdigest_destroy(TDigest *t)
+{
+	pfree(t);
+}
+
+static int
+centroid_cmp(const void *a, const void *b)
+{
+	Centroid *c1 = (Centroid *) a;
+	Centroid *c2 = (Centroid *) b;
+	if (c1->mean < c2->mean)
+		return -1;
+	if (c1->mean > c2->mean)
+		return 1;
+	return 0;
+}
+
+/*
+ * Merges a centroid into the existing list using a scaling function
+ */
+static void
+merge_centroid(mergeArgs *args, const Centroid *merge)
+{
+	float8 k2;
+	Centroid *c = &args->centroids[args->index];
+
+	// integrated_location is used as the scaling function; based on the k_1 scaling function in the
+	// paper k2 corresponds to k(q_right) in the paper
+	args->weight_so_far += merge->weight;
+	k2 = integrated_location(args->t->compression, args->weight_so_far / args->t->total_weight);
+
+	// k1 corresponds to k(q_left) in the paper
+	// if their difference is greater than 1, the resulting merge is not doable
+	// so we move on to the next centroid
+	if (k2 - args->k1 > 1 && c->weight > 0)
+	{
+		// move to the next centroid in the list
+		args->index++;
+		// update the k1 value (qleft) for the next comparison by removing the effect of the
+		// unmerged 'merge' centroid weight
+		args->k1 =
+			integrated_location(args->t->compression,
+								(args->weight_so_far - merge->weight) / args->t->total_weight);
+	}
+
+	// we now merge the centroid at the (possibly updated) index with the incoming
+	// 'merge' centroid
+	c = &args->centroids[args->index];
+	c->weight += merge->weight;
+	c->mean += (merge->mean - c->mean) * merge->weight / c->weight;
+
+	// update the properties of the being-built tdigest
+	if (merge->weight > 0)
+	{
+		args->min = Min(merge->mean, args->min);
+		args->max = Max(merge->mean, args->max);
+	}
+}
+
+/*
+ * Merges the input buffer of centroids into the t-digest's stored centroids
+ * Implements the merge algorithm, which is algorithm 1 of tdunning's paper
+ */
+TDigest *
+ts_tdigest_compress(TDigest *t, List *unmerged_centroids_list)
+{
+	int num_unmerged = list_length(unmerged_centroids_list);
+	Centroid *unmerged_centroids;
+	uint64_t unmerged_weight = 0;
+	ListCell *lc;
+	int i, j;
+	mergeArgs *args;
```

Comment:
can we rename mergeArgs as tdigest_mergeArgs?

Reply:
I would prefer to avoid mixing Camel and low case notations and suggest `tdigest_merge_args`. Camel/Pascal notation is de facto reserved for PG functions and types.

Reply:
Yes, TDigestMergeArgs  sounds good.